### PR TITLE
Do not load calendar/addressbook plugins if not needed

### LINKED ISSUE #6711: Fatal error on synchronizing incorrect vCard file
### Steps to reproduce
1. Upload incorrect vCard file to Nextcloud server (extract [contacts.vcf.zip](https://github.com/nextcloud/server/files/1346735/contacts.vcf.zip))
2. Try to synchronize with desktop client (tested on Windows and Linux with 2.3.2)

### Expected behaviour
File is synchronized

### Actual behaviour
Fatal error on server

### Important concern
The issue is that vCard is parsed. I'll submit an issue in Sabre project about parsing error, but I think Nextcloud shouldn't parse that file in the first place. It should just move files around regardless of their contents.

### General server configuration

**Operating system:** Linux 5a063fefab69 4.4.0-96-generic #119-Ubuntu SMP Tue Sep 12 14:59:54 UTC 2017 x86_64

**Web server:** nginx/1.13.4 (fpm-fcgi)

**Database:** mysql 10.1.26

**PHP version:** 7.0.23
<details>
        <summary>PHP-modules loaded</summary>

```
 - Core
 - date
 - libxml
 - openssl
 - pcre
 - sqlite3
 - zlib
 - ctype
 - curl
 - dom
 - fileinfo
 - filter
 - ftp
 - hash
 - iconv
 - json
 - mbstring
 - SPL
 - PDO
 - bz2
 - posix
 - readline
 - Reflection
 - session
 - SimpleXML
 - pdo_sqlite
 - standard
 - tokenizer
 - xml
 - xmlreader
 - xmlwriter
 - mysqlnd
 - cgi-fcgi
 - apcu
 - bcmath
 - Phar
 - exif
 - gd
 - gettext
 - mcrypt
 - mysqli
 - pdo_mysql
 - shmop
 - sockets
 - sysvmsg
 - sysvsem
 - sysvshm
 - zip
 - memcached
 - Zend OPcache
```
</details>

### Nextcloud configuration

**Nextcloud version:** 12.0.3 - 12.0.3.3

**Updated from an older Nextcloud/ownCloud or fresh install:** Upgrade from 11

**Where did you install Nextcloud from:**  https://github.com/nazar-pc/docker-webserver-apps/tree/master/nextcloud

**Are you using external storage, if yes which one:** files_external is disabled

**Are you using encryption:** no

**Are you using an external user-backend, if yes which one:** No
<details>
        <summary>Signing status</summary>

```
[]
```
</details>

<details>
        <summary>Enabled apps</summary>

```
 - activity: 2.5.2
 - bruteforcesettings: 1.0.2
 - calendar: 1.5.5
 - contacts: 1.5.3
 - dav: 1.3.0
 - federatedfilesharing: 1.2.0
 - files: 1.7.2
 - files_pdfviewer: 1.1.1
 - files_sharing: 1.4.0
 - files_texteditor: 2.4.1
 - files_trashbin: 1.2.0
 - files_versions: 1.5.0
 - files_videoplayer: 1.1.0
 - gallery: 17.0.0
 - issuetemplate: 0.2.2
 - logreader: 2.0.0
 - lookup_server_connector: 1.0.0
 - nextcloud_announcements: 1.1
 - notes: 2.3.1
 - notifications: 2.0.0
 - oauth2: 1.0.5
 - password_policy: 1.2.2
 - provisioning_api: 1.2.0
 - serverinfo: 1.2.0
 - spreed: 2.0.1
 - survey_client: 1.0.0
 - tasks: 0.9.5
 - twofactor_backupcodes: 1.1.1
 - twofactor_totp: 1.3.1
 - updatenotification: 1.2.0
 - workflowengine: 1.2.0
```
</details>

<details>
        <summary>Disabled apps</summary>

```
 - admin_audit
 - comments
 - encryption
 - federation
 - files_external
 - firstrunwizard
 - sharebymail
 - systemtags
 - theming
 - user_external
 - user_ldap
```
</details>

<details>
        <summary>Content of config/config.php</summary>

```
{
    "instanceid": "oc24sob5pzu8",
    "passwordsalt": "***REMOVED SENSITIVE VALUE***",
    "secret": "***REMOVED SENSITIVE VALUE***",
    "trusted_domains": [
        "***REMOVED SENSITIVE VALUE***"
    ],
    "datadirectory": "\/data\/nginx\/www\/data",
    "overwrite.cli.url": "***REMOVED SENSITIVE VALUE***",
    "dbtype": "mysql",
    "version": "12.0.3.3",
    "dbname": "nextcloud",
    "dbhost": "mysql",
    "dbport": "",
    "dbtableprefix": "oc_",
    "dbuser": "***REMOVED SENSITIVE VALUE***",
    "dbpassword": "***REMOVED SENSITIVE VALUE***",
    "logtimezone": "UTC",
    "installed": true,
    "maintenance": false,
    "theme": "",
    "loglevel": 2,
    "updater.release.channel": "stable"
}
```
</details>

### Logs

<details>
        <summary>Nextcloud log (data/nextcloud.log)</summary>

```
Sabre\VObject\ParseException: Invalid Mimedir file. Line starting at 13 did not follow iCalendar/vCard conventions

    /data/nginx/www/3rdparty/sabre/vobject/lib/Parser/MimeDir.php - line 234: Sabre\VObject\Parser\MimeDir->readProperty('=0ABaytown\\, LA...')
    /data/nginx/www/3rdparty/sabre/vobject/lib/Parser/MimeDir.php - line 181: Sabre\VObject\Parser\MimeDir->parseLine('=0ABaytown\\, LA...')
    /data/nginx/www/3rdparty/sabre/vobject/lib/Parser/MimeDir.php - line 89: Sabre\VObject\Parser\MimeDir->parseDocument()
    /data/nginx/www/3rdparty/sabre/vobject/lib/Reader.php - line 46: Sabre\VObject\Parser\MimeDir->parse('BEGIN VCARD\r\nVE...', 0)
    /data/nginx/www/3rdparty/sabre/dav/lib/CardDAV/Plugin.php - line 857: Sabre\VObject\Reader read('BEGIN VCARD\r\nVE...')
    /data/nginx/www/3rdparty/sabre/dav/lib/CardDAV/Plugin.php - line 782: Sabre\CardDAV\Plugin->convertVCard('BEGIN VCARD\r\nVE...', 'vcard3')
    [internal function] Sabre\CardDAV\Plugin->httpAfterGet(Object(Sabre\HTTP\Request), Object(Sabre\HTTP\Response))
    /data/nginx/www/3rdparty/sabre/event/lib/EventEmitterTrait.php - line 105: call_user_func_array(Array, Array)
    /data/nginx/www/3rdparty/sabre/dav/lib/DAV/Server.php - line 491: Sabre\Event\EventEmitter->emit('afterMethod GET', Array)
    /data/nginx/www/3rdparty/sabre/dav/lib/DAV/Server.php - line 254: Sabre\DAV\Server->invokeMethod(Object(Sabre\HTTP\Request), Object(Sabre\HTTP\Response))
    /data/nginx/www/apps/dav/lib/Server.php - line 258: Sabre\DAV\Server->exec()
    /data/nginx/www/apps/dav/appinfo/v2/remote.php - line 33: OCA\DAV\Server->exec()
    /data/nginx/www/remote.php - line 162: require_once('/data/nginx/www...')
    {main}
```
</details>

### PR DESCRIPTION
**Pull request by rullzer**
Fixes #6711
Based on https://github.com/owncloud/core/pull/30149/commits/f0fb21cf3563088284c56be4eac48c7d9725cb21

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

### CODE REVIEW

**Comment by MorrisJobke** [comment]
Does this also fix stuff like #8141?

**Comment by georgehrke** [comment]
@MorrisJobke It should be as far as i can tell.

**Comment by georgehrke** [comment]
I will check when testing this pr.

**Comment by rullzer** [comment]
Yes it should

**Review by MorrisJobke — APPROVED** [review]
Tested, works and code makes sense 👍

**Comment by MorrisJobke** [comment]
CI failure is due to a timeout :)

**Comment by rullzer** [comment]
One more review please :)

**Comment by rullzer** [comment]
Conflicts fixed and rebased.

**Comment by MorrisJobke** [comment]
@georgehrke @blizzz @nickvergessen @ChristophWurst @skjnldsv Reviews would be nice :)

**Review by blizzz — CHANGES_REQUESTED** [review]
the repair step might lose valid shares when

1. Nextcloud is in maintenance mode
2. AND external user backends are used

since apps are not loaded in this context.

**Comment by rullzer** [comment]
@blizzz I can drop the repair job... but then we might have invalid shares....

**Comment by blizzz** [comment]
> @blizzz I can drop the repair job... but then we might have invalid shares....

Or add a switch that prevents running in maintenance mode.

**Comment by rullzer** [comment]
@blizzz now they are only run if you run the expensive repair steps in case you run into it.

All our other repair steps are executed in maintenance mode so then it indeed would do :boom: in certain situations

**Comment by blizzz** [comment]
> All our other repair steps are executed in maintenance mode so then it indeed would do :boom: in certain situations

Then it is better run as a seperate occ command, bound to apps, as i am afraid it would do more harm than good.

**Comment by rullzer** [comment]
Nah during upgrade.

Now it is. occ maintenance:repair - - expensive

**Comment by rullzer** [comment]
@blizzz ok one more time ;)

**Review by blizzz — APPROVED** [review]

**Comment by blizzz** [comment]
looks good now, apart from tests

File: lib/private/Repair.php [hunk 1]
```diff
@@ -47,6 +47,8 @@
 use OC\Repair\RepairInvalidShares;
 use OC\Template\JSCombiner;
 use OC\Template\SCSSCacher;
+use OCA\DAV\Connector\Sabre\Principal;
+use OCA\DAV\Repair\RemoveInvalidShares;
```
Comment (MorrisJobke) [inline lib/private/Repair.php]: bildschirmfoto 2018-05-31 um 17 22 21

🙈

**Pull request merged**